src/Draco.Compiler/Internal/Semantics/AbstractSyntaxTree.cs

```diff
@@ -0,0 +1,196 @@
+    /// </summary>
+    public partial record class FuncBody : AbstractSyntaxTree
+    {
+        /// <summary>
+        /// A block function body.
+        /// </summary>
+        public sealed partial record class BlockBody(
+            Expr.Block Block) : FuncBody;
+
+        /// <summary>
+        /// An in-line function body.
+        /// </summary>
+        public sealed partial record class InlineBody(
+            Expr Expression) : FuncBody;
+    }
+
+    public abstract partial record class Expr : AbstractSyntaxTree
+    {
+        /// <summary>
+        /// A block expression
+        /// </summary>
+        public partial record class Block(
+            ImmutableArray<Stmt> Statements,
+            Expr? Value) : Expr;
+
+        /// <summary>
+        /// A literal expression, i.e. a number, string, boolean value, etc.
+        /// </summary>
+        public sealed partial record class Literal(
+            Symbol Value) : Expr;
+
+        /// <summary>
+        /// An if-expression with an option else clause.
+        /// </summary>
+        public sealed partial record class If(
+            Expr Condition,
+            Expr Then,
+            Expr? Else) : Expr;
+
+        /// <summary>
+        /// A while-expression.
+        /// </summary>
+        public sealed partial record class While(
+            Expr Condition,
+            Expr Expression) : Expr;
+
+        /// <summary>
+        /// A goto-expression.
+        /// </summary>
+        public sealed partial record class Goto(
+            Symbol Target) : Expr;
+
+        /// <summary>
+        /// A return-expression.
+        /// </summary>
+        public sealed partial record class Return(
+            Expr? Expression) : Expr;
+
+        /// <summary>
+        /// Any call expression.
+        /// </summary>
+        public sealed partial record class Call(
+            Expr Called,
+            ImmutableArray<Expr> Args) : Expr;
+
+        /// <summary>
+        /// Any index expression.
+        /// </summary>
+        public sealed partial record class Index(
+            Expr Called,
+            ImmutableArray<Expr> Args) : Expr;
+
+        /// <summary>
+        /// A member access expression.
+        /// </summary>
+        public sealed partial record class MemberAccess(
+            Expr Object,
+            Symbol MemberName) : Expr;
+
+        /// <summary>
+        /// A unary expression.
+        /// </summary>
+        public sealed partial record class Unary(
+            Symbol Operator,
+            Expr Operand) : Expr;
+
+        /// <summary>
+        /// A binary expression, including assignment and compound assignment.
+        /// </summary>
+        public sealed partial record class Binary(
+            Expr Left,
+            Symbol Operator,
+            Expr Right) : Expr;
+
+        /// <summary>
+        /// A relational expression chain.
+        /// </summary>
+        public sealed partial record class Relational(
```

Comment:
This is tricky, we should discuss what to transform this into to avoid complex codegen